apple/internal/xcframework_rules.bzl

```diff
@@ -0,0 +1,675 @@
+    """Generates the bundle archive for an XCFramework.
+
+     Args:
+        actions: The actions providerx from `ctx.actions`.
+        bundle_name: The name of the XCFramework bundle.
+        framework_archive_files: A list of depsets referencing files to be used as inputs to the
+            bundling action. This should include every archive referenced as a "src" of
+            framework_archive_merge_zips.
+        framework_archive_merge_zips: A list of structs representing ZIP archives whose contents
+            should be merged into the bundle. Each struct contains two fields: "src", the path of
+            the archive whose contents should be merged into the bundle; and "dest", the path inside
+            the bundle where the ZIPs contents should be placed. The destination path is relative to
+            `bundle_path`.
+        label_name: Name of the target being built.
+        output_archive: The file representing the final bundled archive.
+        resolved_bundletool: A struct referencing the resolved bundle tool.
+        root_info_plist: A `File` representing a fully formed root Info.plist for this XCFramework.
+    """
+    bundletool_control_file = intermediates.file(
+        actions = actions,
+        target_name = label_name,
+        output_discriminator = None,
+        file_name = "xcframework_bundletool_control.json",
+    )
+    bundletool_control = struct(
+        bundle_merge_files = [struct(src = root_info_plist.path, dest = "Info.plist")],
+        bundle_merge_zips = framework_archive_merge_zips,
+        bundle_path = bundle_name + ".xcframework",
+        output = output_archive.path,
+    )
+    actions.write(
+        output = bundletool_control_file,
+        content = json.encode(bundletool_control),
+    )
+
+    actions.run(
+        arguments = [bundletool_control_file.path],
+        executable = resolved_bundletool.executable,
+        inputs = depset([bundletool_control_file, root_info_plist], transitive = [resolved_bundletool.inputs] + framework_archive_files),
+        input_manifests = resolved_bundletool.input_manifests,
+        mnemonic = "CreateXCFrameworkBundle",
+        outputs = [output_archive],
+        progress_message = "Bundling %s" % label_name,
+    )
+
+def _apple_xcframework_impl(ctx):
+    """Experimental WIP implementation of apple_xcframework."""
+
+    actions = ctx.actions
+    apple_toolchain_info = ctx.attr._toolchain[AppleSupportToolchainInfo]
+
+    # Bundle extension needs to be ".xcframework" for root bundle, but macos/ios/tvos will always
+    # be ".framework"
+    nested_bundle_extension = ".framework"
+
+    # Similarly, bundle_id is expected to be in terms of the bundle ID for each embedded framework,
+    # as this value is not used in the XCFramework's root Info.plist.
+    nested_bundle_id = ctx.attr.bundle_id
+    bundle_name = ctx.attr.bundle_name or ctx.attr.name
+    executable_name = getattr(ctx.attr, "executable_name", bundle_name)
```

Comment:
@thii I didn't use `bundling_support.executable_name()` as it didn't play nice with this specific rule. We can probably fix it up in a followup PR.